articles/WideStrings.md

```diff
@@ -0,0 +1,213 @@
+Other subscribers like `ros2 topic echo` may echo the bytes in hexadecimal.
+
+In practice using different encodings may not be possible.
+The IDL specification forbids `string` from containing `NULL` values.
+For compatibility a ROS message `string` field must not contain zero bytes, and a `wstring` field must not contain zero words.
+An encoding like UTF-32 cannot be used since it may have zero bytes or words in a code point.
+If UTF-32 encoded strings need to be transmitted then one should either convert to UTF-8 and use a `string` or use a `uint32` array.
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> Can ROS 1 publish a string with NULL bytes?
+</div>
+
+## Unicode Strings Across ROS 1 Bridge
+Since ROS 1 and 2 both allow `string` to be UTF-8, the ROS 1 bridge will pass values unmodified between them.
+If a message sent from ROS 2 to ROS 1 fails to serialize because a string contains values that are not legal UTF-8 then it won't be published by the bridge.
+
+If a ROS 2 message has a field of type `wstring` then the bridge will attempt to convert it from UTF-16 to UTF-8.
+The resulting UTF-8 encoded string will be published as a `string` type.
+If the conversion fails then the bridge will not publish the message.
+
+## Size of a Wide String
+
+Both UTF-8 and UTF-16 are variable width encodings.
+To minimize the amount of memory used, the `string` and `wstring` types are to be stored in client libraries according to the smallest possible code point.
+This means `string` must be specified as a sequence of bytes, and `wstring` is to be specified as a sequence of words.
+
+Some DDS implementations use 32bit types to store wide strings values.
+This may be due to DDS-XTYPES 1.1 spec saying `wchar` is a 32bit value.
+ROS 2 will instead aim to be compatible with DDS-XTYPES 1.2 and use 16bit storage for wide characters.
+Generated code for ROS 2 messages will automatically handle the conversion when a message is serialized or deserialized.
+
+### Bounded wide strings
+
+Embedded systems accepting strings may use message definitions that restrict the maximum size of a string.
+These are referred to as bounded strings.
+Their purpose is to restrinct the amount of memory used, so the bounds must be specified as units of memory.
+If a `string` field is bounded then the size is given in bytes.
+Similarly the size of a bounded `wstring` is to be specified in words.
+It is the responsibility of whoever populates a bounded `string` or `wstring` to make sure it contains whole code points only.
+Partial code points are indistinuguishable from unknown encodings, so a bounded string whose last code point is incomplete will be published without error.
+
+## Runtime impact of wide string
+
+Dealing with wide strings puts more strain on the software of a system, both in terms of speed and code size.
+UTF-8 and UTF-16 are both variable width encodings, meaning a code point can take 1 (UTF-8 only), 2, or 4 bytes to represent.
```

Comment:
In the case of UTF-8 3 bytes is also an option.

Reply:
ddf99014af761f0965d54086b1b49f8fb420eb4a